pkg/cloudprovider/providers/gce/gce_disks.go

```diff
@@ -727,21 +769,50 @@ func getDiskType(diskType string) (string, error) {
 	}
 }
 
-func (gce *GCECloud) DeleteDisk(diskToDelete string) error {
-	err := gce.doDeleteDisk(diskToDelete)
+func (gce *GCECloud) DeleteDisk(name string, zone string) error {
+	if zone == "" {
+		return fmt.Errorf("a zone must be specified")
```

Comment:
This change requires that the PV must have the zone label on it?  Did it previously work with PVs that didn't? Trying to make sure we don't cause a regression

Reply:
Previously it didn't depend on PV labels at all. It used to search through all zones for a PD with the given name and deletes it regardless of whether it was in the right zone.

I don't view it as a regression - deleting a disk without zone information is unsafe, so IMO it's better to return an error.

Reply:
I agree it's safer and better, but if there is a possibility that there are PVs out there that don't have a zone label on it, then it could leak pds (if users don't check that the pv actually got deleted).  With the PV label admission controller, the likelihood is lower, however, we didn't always have the admission controller enabled.

Reply:
Although it didn't used to leak PDs in the case you described, the previous behavior could be considered a bug actually (rather than the proposed behavior being a regression), since it's making potentially incorrect assumptions. Essentially we have to make a trade off between the following possibilities:
* Missing zone label, causing leaked PDs (and undeleted PVs with error event).
* Storage admin manually deleting a PD, creating a PD with the same name in a different zone, and a user deleting a PVC bound to the PV for the first PD. In this case the wrong disk is deleted.

The latter option is arguably far less likely, but the penalty is greater. I'll leave this open for discussion for now

Reply:
There's also just the case where a user manually creates a PV without specifying the zone label, and didn't have the admission controller running (in GKE, we didn't enable the admission controller until 1.6).  In that case, the PV deletion will now fail even though everything else is fine (ie, there's no other conflicting disk).  

As a safeguard, can we detect if there are multiple disks with the same name, and only fail deletion in that case?

Reply:
That's the previous behavior, so the problem only really comes up when there's a new PD with the same name but different zone, *and* the previous PD is gone.

Reply:
Discussed with @msau42 offline. A similar error could crop up when a user deletes the old PD and creates a new one with the same name *and the same zone*, so even with the zone label there isn't enough information here to decide whether the disk is safe to delete.

Since not deleting the disk doesn't really solve this problem entirely, and this case is very unlikely to happen, I'm leaning towards moving back to the old behavior.

Reply:
The truly correct way to solve this issue is to have something that uniquely identifies every instance of a disk, even over time (i.e. disks with the same name and zone but exists at different times should have different identifiers).

Reply:
GCE PD actually do have a unique identifier like you mentioned: https://cloud.google.com/compute/docs/reference/rest/v1/disks/insert (See "id" field)
Maybe we can incorporate that in identification somehow.

In terms of backwards compatibility I do agree that deleting the wrong disk is a much bigger potential problem than not deleting the underspecified right disk. So if it is a choice between the two I would lean towards the latter.

Reply:
Nice! I see the Id field in the Go client too. By the way @saad-ali the disk resource has a `selfLink` which might be the fully qualified URL you were referring to a while back, but it only contains project, zone, and disk name information.

Reply:
@jingxu97 @saad-ali interested in your opinions here too.